doc/Container Hardening/SONiC_container_hardening_HLD.md

```diff
@@ -0,0 +1,347 @@
+		- [6.2. CLI/YANG model Enhancements](#62-cliyang-model-enhancements)
+		- [6.3. Config DB Enhancements](#63-config-db-enhancements)
+	- [7. Warmboot and Fastboot Design Impact](#7-warmboot-and-fastboot-design-impact)
+	- [8. Restrictions/Limitations](#8-restrictionslimitations)
+	- [9. Testing Requirements/Design](#9-testing-requirementsdesign)
+		- [9.1 Unit Test cases](#91-unit-test-cases)
+		- [9.2 System Test cases](#92-system-test-cases)
+	- [10. Open/Action items - if any](#10-openaction-items---if-any)
+	- [Appendix A: Further reading](#appendix-a-further-reading)
+	- [Appendix B: Linux Capabilities](#appendix-b-linux-capabilities)
+
+## List of Tables
+* [Table 1: Revision](#table-1-revision)
+* [Table 2: Abbreviations](#table-2-abbreviations)
+* [Table 3: Default Linux capabilities](#table-3-default-linux-capabilities)
+* [Table 4: Extended Linux capabilities](#table-4-extended-linux-capabilities)
+
+## Revision
+###### Table 1: Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 |             |                    | Initial version                   |
+
+## Scope
+
+This section describes the requirements, goals, and recommendations of the container hardening item for SONiC.
+
+## Definitions/Abbreviations
+###### Table 2: Abbreviations
+| Definitions/Abbreviation | Description                                |
+|--------------------------|--------------------------------------------|
+| OS                       | Operating System                           |
+| API                      | Application Programmable Interface         |
+| SAI                      | Swich Abstraction Interface                |
+
+## 1. Overview
+
+Containers is a method of creating virtualization and abstraction of an OS for a subset of processes/service on top of a single host with the purpose of giving it an environment to run and execute its tasks without effect of nearby containers/processes.
+
+In SONiC, we are deploying containers with full visibility and capabilities as the host Linux.
+
+This poses a security risk and vulnerability as a single breached container means that the whole system is breached.
+
+Addressing this issue – we have composed this doc for container hardening, describing the security hardening requirements and definitions for all containers on top of SONiC.
+
+## 2. Requirements
+
+What are we trying to achieve here?
+
+We would like to increase the security in SONiC so that an attack on a specific container will not compromise the whole system.
+
+To do so, we'll tackle the following areas:
+1. Privileges
+2. Network
+3. Capabilities
+4. Mount namespace
+5. Cgroups
+6. Etc
+
+For now, we will focus on #1 & #2
+
+Further guidelines and requirements will be brought upon in the future on-demand.
+
+## 3. Architecture Design
+
+### 3.1 Root privileges
```

Comment:
>Root privileges [](http://example.com/codeflow?start=8&length=15)

I'm confused, how removing `--privileged` flag can also remove root privileged?

Reply:
The `--privileged` flag for the container provides it with all Linux capabilities (definition of root).
If we remove it - we'll get a basic set of system capabilities.

BTW - additional steps can be to further reduce the Linux capabilities given to a container to a more fine-grained requirements so that each container will only be able to perform a specific set of actions that it requires.